Move corefx onto the .NET Core 2.0 toolset

mellinoe opened the pull request:
The latest version of buildtools allows a 2.0 CLI to be used. This change pulls in that version of buildtools and updates corefx to use the 2.0 CLI for restoring nuget packages and running our shared tools.

* All of the project.json(.template) files under `external` have been converted to use `PackageReference` items directly defined in the depproj files. The versions for these references are based on pre-existing properties defined in dependencies.props. When we take future version update PR's, we should only see changes in that single file.
* packageresolve.targets was modified so that the paths for the MSBuild project to be restored, as well as the "project.assets.json" file (the lock file, essentially). Note that the property for the projec to be restored is still called `ProjectLockJson`, although it represents an MSBuild project. I'll look into cleaning this up, I just need to make sure there's no stray references to "ProjectLockJson" remaining.
* There are a bunch of new calls to `chmod` here and there. This is because .NET Core 2.0 (which we now restore all packages with, and run the build on) does not preserve executable flags in Unix files, by design. We need to mark our programs as executable before we try to use them.
* NuGet now restores all packages into folders with lower-case names. This needed to be accounted for in a variety of different places.
* ~~I've brought back the "hack" where we pre-build CoreFx.Tools.dll before running the full product build. This is necessary due to changes in the CLI. Technically this will go away when we move CoreFx.Tools.dll into buildtools.~~ Never mind. We don't need that.

NOTE: This is not going to work on Alpine right now, because there is no 2.0 CLI being produced for it. Everything should be in place to do, so, though, so hopefully we can turn that on shortly.

@ericstj , @weshaggard, @joperezr

joperezr commented on build.sh:
```diff
@@ -25,7 +25,13 @@ __scriptpath=$(cd "$(dirname "$0")"; pwd -P)
+fi
+
+# Building CoreFx.Tools before calling build-managed.sh to ensure our tools are available before running the main build.
+"$__scriptpath/Tools/msbuild.sh" "$__scriptpath/src/Tools/CoreFx.Tools/CoreFx.Tools.csproj" /v:m /m
```
Just curious, Did we end up figuring out what changed on 2.0 in the last month that regressed this?
Also, don't we need to do this as well on build.cmd?

mellinoe replied:
Actually, this was a misunderstanding on my part. There was a configuration option in MSBuild.runtimeconfig.json (which we override) that gave us the old behavior back. I accidentally deleted that file. After un-deleting it, we don't need this hacky stuff.

joperezr commented on dependencies.props:
```diff
@@ -171,11 +172,5 @@
     </DependencyBuildInfo>
-
-    <!-- project.json files to update -->
-    <ProjectJsonFiles Include="$(MSBuildThisFileDirectory)external\**\project.json" />
```
I assume that `msbuild /t:UpdateDependencies` will still work?

mellinoe replied:
Yep, it does. It will only update dependencies.props now, but the properties defined in there are used in the depproj files.

joperezr commented on dir.props:
```diff
@@ -104,7 +104,7 @@
     <DnuRestoreCommand Condition="'$(ParallelRestore)'=='true'">$(DnuRestoreCommand) --parallel</DnuRestoreCommand>
     <DnuRestoreCommand Condition="'$(UseNuGetHttpCache)'!='true'">$(DnuRestoreCommand) --no-cache</DnuRestoreCommand>
-    <DnuRestoreCommand>$(DnuRestoreCommand) --packages "$(PackagesDir.TrimEnd('/\'.ToCharArray()))" $(DnuRestoreSource)</DnuRestoreCommand>
+    <DnuRestoreCommand>$(DnuRestoreCommand) --packages "$(PackagesDir.TrimEnd('/').TrimEnd('\'))" $(DnuRestoreSource)</DnuRestoreCommand>
```
mm not sure I understood this change.

mellinoe replied:
For some reason, MSBuild did not want to call the `TrimEnd` overload we were using before. We did add a parameterless overload to this function in 2.0. I couldn't figure out the syntax to get it right, so I just made to calls to TrimEnd instead.

eerhardt replied:
There was some breaking changes in MSBuild with netcoreapp2.0 caused by the new overload.  See [https://github.com/Microsoft/msbuild/issues/1634](https://github.com/Microsoft/msbuild/issues/1634).

/cc @rainersigwald @AndyGerlicher - in case they want to know about this change for netcoreapp2.0.  It feels like a bug.

---
In reply to: [110289282](https://github.com/dotnet/corefx/pull/18035#discussion_r110289282) [](ancestors = 110289282)

rainersigwald replied:
Filed https://github.com/Microsoft/msbuild/issues/1966 to take a look at this

joperezr commented on external/dir.props:
```diff
@@ -8,5 +8,8 @@
     <NugetRuntimeIdentifier>$(RuntimeOS)-$(ArchGroup)</NugetRuntimeIdentifier>
+    <ContainsPackageReferences>true</ContainsPackageReferences>
+    <!-- We need configuration-specific assets files. -->
+    <RestoreOutputPath>$(BaseIntermediateOutputPath)external/$(MSBuildProjectName)/$(TargetGroup)</RestoreOutputPath>
```
I believe instead of TargetGroup, you could use `$(Configuration)` instead

weshaggard replied:
Better yet just hook it off IntermedateOutputPath (https://github.com/dotnet/corefx/blob/master/dir.props#L237)

mellinoe replied:
> Better yet just hook it off IntermedateOutputPath 

Yeah that seems simple enough. Will change.

joperezr commented on external/ilasm/ilasm.depproj:
```diff
@@ -6,6 +6,22 @@
   </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NETCore.Platforms">
+      <Version>2.0.0-$(CoreFxExpectedPrerelease)</Version>
```
I have a feeling that this will eventually break us in the future. Instead we might consider using the actual nuget package version in a property, instead of just the prerelease. This is doable through dependencies.props. cc: @davisgoodin

mellinoe replied:
Agreed, that would be better. I'm not very familiar with the format of dependencies.props, but if we can get that to work then I'm all for it.

weshaggard replied:
I imagine we will do a clean-up pass of a lot of these versions once we get fully moved to the msbuild based CLI.

joperezr commented on external/ilasm/ilasm.depproj:
```diff
@@ -6,6 +6,22 @@
     <NugetRuntimeIdentifier>$(RuntimeOS)-x64</NugetRuntimeIdentifier>
```
>RuntimeOS [](start = 30, length = 9)

Where are you specifying the TFM of these guys now?

mellinoe replied:
That's still getting pulled in through the build configuration, unless I'm misunderstanding the question.

weshaggard replied:
Yes it should be.

weshaggard commented on Tools-Override/crossgen.sh:
```diff
@@ -20,15 +22,16 @@ restore_crossgen()
-    echo "{\"frameworks\":{\"netcoreapp1.1\":{\"dependencies\":{\"Microsoft.NETCore.Runtime.CoreCLR\":\"$__CoreClrVersion\", \"Microsoft.NETCore.Platforms\": \"$__CoreClrVersion\"}}},\"runtimes\":{\"$__rid\":{}}}" > "$__pjDir/project.json"
-    $__dotnet restore $__pjDir/project.json --packages $__packagesDir
-    __crossgenInPackage=$__packagesDir/runtime.$__packageRid.Microsoft.NETCore.Runtime.CoreCLR/$__CoreClrVersion/tools/crossgen
+    echo "<Project Sdk=\"Microsoft.NET.Sdk\"><PropertyGroup><TargetFramework>netcoreapp2.0</TargetFramework><DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences><RuntimeIdentifiers>$__packageRid</RuntimeIdentifiers></PropertyGroup><ItemGroup><PackageReference Include=\"Microsoft.NETCore.Runtime.CoreCLR\" Version=\"$__CoreClrVersion\" /><PackageReference Include=\"Microsoft ...
```
Do we need this any more? Has crossgen been added to the package yet?

mellinoe replied:
It wasn't when I was using an "older" version of the 2.0 CLI. Let me check again now that this works on a newer version.

weshaggard commented on Tools-Override/crossgen.sh:
```diff
@@ -2,7 +2,9 @@
 # Restores crossgen and runs it on all tools components.
 
-__CoreClrVersion=1.1.0
+__CoreClrVersion=2.0.0-preview1-25204-02
```
The duplication of all these version is going to be a maintenance nightmare we need to think about ways of eliminating them.

mellinoe replied:
Technically I can pluck this version out of the string resources of libcoreclr.so. If `crossgen` is included with the shared framework though (which I think is the plan?) then it won't be necessary.

ellismg replied:
> If crossgen is included with the shared framework though (which I think is the plan?).

It should now be in the the tools folder of the Microsoft.NETCore.App package (so you don't need to restore the Microsoft.NETCore.Runtime.CoreCLR package) but the current plan is to not include it as an exe in the shared framework on disk.

/cc @gkhanna79

weshaggard commented on external/ilasm/ilasm.depproj:
```diff
@@ -6,6 +6,22 @@
+  </ItemGroup>
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.targets))\dir.targets" />
+  <!-- When zipped/unzipped in a package, ilasm loses its executable flag. Reset it. -->
+  <Target Name="MakeIlasmExecutable" AfterTargets="CopyFilesToOutputDirectory" Condition="'$(RunningOnUnix)' == 'true'">
```
Why wasn't this an issue before?

mellinoe replied:
See this bullet point:

> * There are a bunch of new calls to chmod here and there. This is because .NET Core 2.0 (which we now restore all packages with, and run the build on) does not preserve executable flags in Unix files, by design. We need to mark our programs as executable before we try to use them.

weshaggard commented on external/dir.targets:
```diff
@@ -1,33 +1,11 @@
   <Import Project="..\dir.targets" />
 
   <PropertyGroup>
     <RestorePackages>true</RestorePackages>
```
Are these properties still needed?

mellinoe replied:
I believe so... `RestorePackages` defaults to false and is only set in this folder.

weshaggard replied:
I guess it won't be needed once we stop doing the restore ourselves and start using the nuget targets for restore.

weshaggard commented on external/harvestPackages/harvestPackages.netstandard1.6.depproj:
```diff
@@ -0,0 +1,143 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.CSharp">
```
I assume you are keeping the Version as nested metadata because of the dependency on a new msbuild we don't want to take just yet?

mellinoe replied:
Yes, that's right. It would prevent us from using VS 2015 on Windows, and the only benefit we'd get would be that syntax sugar.

weshaggard commented on external/netfx/netfx.depproj:
```diff
@@ -13,10 +13,16 @@
   <PropertyGroup Condition="!$(NuGetTargetMoniker.StartsWith('.NETFramework'))">
     <!-- For things not .NETFramework we need the net461 targeting pack to generate facades -->
     <NuGetTargetMoniker>.NETFramework,Version=v4.6.1</NuGetTargetMoniker>
+    <NuGetTargetMonikerShort>net461</NuGetTargetMonikerShort>
```
Were does NugetTargetMonikerShort get consumed?

mellinoe replied:
It gets consumed here in [packageresolve.targets](https://github.com/dotnet/corefx/pull/18035/files#diff-3045b99d7cac6dab2d9cddaf1b3cbc2fR40). Unfortunately, passing the "full moniker" to `/p:TargetFramework` does not work with the 2.0 CLI / SDK.

weshaggard commented on external/runtime/runtime.depproj:
```diff
@@ -7,18 +7,45 @@
     <NugetRuntimeIdentifier Condition="'$(OSGroup)' == 'Unix' AND '$(RunningOnUnix)' != 'true'">ubuntu.14.04-x64</NugetRuntimeIdentifier>
     <NugetRuntimeIdentifier Condition="'$(TargetGroup)' == 'uap'">win10-$(ArchGroup)</NugetRuntimeIdentifier>
     <NugetRuntimeIdentifier Condition="'$(TargetGroup)' == 'uapaot'">win10-$(ArchGroup)-aot</NugetRuntimeIdentifier>
+    <AppendMSBuildRestoreArguments>true</AppendMSBuildRestoreArguments>
```
What is AppendMSBuildRestoreArguments and why do we need it?

mellinoe replied:
Oops, I think this is a leftover change from some intermediate state. Will remove.

weshaggard commented on external/runtime/runtime.depproj:
```diff
@@ -7,18 +7,45 @@
+    </PackageReference>
+  </ItemGroup>
+
+
```
nit: extra line

weshaggard commented on init-tools.cmd:
```diff
@@ -10,8 +10,8 @@ if [%BUILDTOOLS_SOURCE%]==[] set BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/do
-set PROJECT_JSON_FILE=%PROJECT_JSON_PATH%\project.json
-set PROJECT_JSON_CONTENTS={ "dependencies": { "Microsoft.DotNet.BuildTools": "%BUILDTOOLS_VERSION%" }, "frameworks": { "netcoreapp1.0": { } } }
+set PROJECT_JSON_FILE=%PROJECT_JSON_PATH%\project.csproj
+set PROJECT_JSON_CONTENTS=^^^<Project Sdk=^^^"Microsoft.NET.Sdk^^^"^^^>^^^<PropertyGroup^^^>^^^<TargetFramework^^^>netcoreapp1.0^^^</TargetFramework^^^>^^^<DisableImplicitFrameworkReferences^^^>true^^^</DisableImplicitFrameworkReferences^^^>^^^</PropertyGroup^^^>^^^<ItemGroup^^^>^^^<PackageReference Include=^^^"Microsoft.DotNet.BuildTools^^^" Version=^^^"%BUILDTOOLS_VERSION%^^^" /^^^>^^^</ItemGro ...
```
Similar comment (with similar response I'm sure :)) - This should be a file now that we can pass parameters.

mellinoe replied:
I can do that. Any opinions on what that project file should be called / where it should be?

weshaggard replied:
I think I would just call it something like init-tools.proj (or perhaps init-tools.msbuild to avoid clashing with other projs) and put it right next to the init-tools scripts.

weshaggard commented on init-tools.cmd:
```diff
@@ -69,14 +69,6 @@ if not [%INIT_TOOLS_ERRORLEVEL%]==[0] (
   exit /b %INIT_TOOLS_ERRORLEVEL%
 )
 
-echo Updating CLI NuGet Frameworks map...
```
Yay!

weshaggard commented on init-tools.sh:
```diff
@@ -144,24 +145,22 @@ if [ ! -e $__INIT_TOOLS_DONE_MARKER ]; then
+    echo "Making all .sh files executable under Tools."
+    ls $__scriptpath/Tools/*.sh | xargs chmod +x
+
+    cp Tools-Override/* Tools
```
This is currently being done in run.sh https://github.com/dotnet/corefx/blob/master/run.sh#L10.

weshaggard commented on src/Tools/CoreFx.Tools/CoreFx.Tools.builds:
```diff
@@ -5,7 +5,6 @@
     <Project Include="$(MSBuildThisFileDirectory)$(MSBuildProjectName).csproj" />
     <Project Include="$(MSBuildThisFileDirectory)$(MSBuildProjectName).csproj">
```
Given the way we build this https://github.com/dotnet/corefx/blob/master/build.proj#L41, this entire file can probably be deleted.

weshaggard commented on src/Tools/GenerateProps/targetgroups.props:
```diff
@@ -4,181 +4,212 @@
     <TargetGroups Include="netcore50">
       <TargetingPackNugetPackageId>Microsoft.TargetingPack.Private.NetNative</TargetingPackNugetPackageId>
       <NuGetTargetMoniker>.NETCore,Version=v5.0</NuGetTargetMoniker>
+      <NuGetTargetMonikerShort>netcore50</NuGetTargetMonikerShort>
```
Given you have these here why do you have them in all the project files as well?

mellinoe replied:
I only put them in the project files where the TFM was manually overridden.

weshaggard commented on src/Tools/corefxTools.props:
```diff
@@ -5,7 +5,7 @@
     <!-- Temporarily enable local build of tools -->
     <CoreFxToolsDir Condition="'$(CoreFxToolsDir)' == ''">$(ToolsDir)</CoreFxToolsDir>
     <CoreFxDesktopToolsDir Condition="'$(CoreFxDesktopToolsDir)' == ''">$(ToolsDir)net46/</CoreFxDesktopToolsDir>
-    <CoreFxToolsTaskDir Condition="'$(CoreFxToolsTaskDir)' == '' AND '$(RunningOnCore)' == 'true'">$(CoreFxToolsDir)</CoreFxToolsTaskDir>
```
What happen to our RunningOnCore property?

mellinoe replied:
I'm not really sure why I changed this. I'll look into it.

weshaggard commented on src/src.builds:
```diff
@@ -12,7 +12,7 @@
   <PropertyGroup>
     <!-- TODO: We should see about generating this from scratch instead of relying on a previous deps file as a template -->
-    <_ToolsDotNetCliSharedFxPath>$(ToolsDir)dotnetcli\shared\Microsoft.NETCore.App\1.1.0</_ToolsDotNetCliSharedFxPath>
+    <_ToolsDotNetCliSharedFxPath>$(ToolsDir)dotnetcli\shared\Microsoft.NETCore.App\1.1.1</_ToolsDotNetCliSharedFxPath>
```
Why isn't this 2.0.0 now?

mellinoe replied:
Oops, I believe it should be. This is another annoying place where versions are hardcoded...

mellinoe commented:
Hmm, it looks like something is broken on a lot of the Unix builds. I logged into the CentOS7.1 machine and it turns out that `build.sh` works for me, but `./build.sh -release -framework:netcoreapp -os:Linux -buildArch:x64` (what the CI passes) doesn't. I'll take a look at why it's not working.

mellinoe commented:
I fixed the way some parameters were getting passed around. Hopefully this will fix those issues.

mellinoe commented:
Ok, I'm hopeful that this next build will be successful. I pushed another update that fixed a problem caused by the new "mono" TargetGroup not having a NuGetTargetMonikerShort value.

mellinoe commented:
@karajas Can you look at the latest commit? I added some logic to GenerateDepsFile which filters out items from the "native" section of the file when they aren't contained in our test folder. I was hitting errors because on Windows, the deps file listed "DiaSymReader.dll" when we don't have that in our test folder. There were a handful of such files that are filtered out and removed with my new logic.

weshaggard commented on external/ILLink/illink.runtimeconfig.json:
```diff
@@ -0,0 +1,9 @@
+{
+  "runtimeOptions": {
+    "tfm": "netcoreapp1.1",
```
What is the tfm option? I'm not sure I've seen that before.

mellinoe replied:
I'm not sure. It is in the runtimeconfig.json file included in the ILLink nuget package, and this file is copy-pasted from there. I just updated the Microsoft.NETCore.App version below.

eerhardt replied:
The new tooling puts this "tfm" in the runtimeconfig.  This value is used by the host to do "shared runtime store" probing.  See [https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/DotNetCore-SharedPackageStore.md](https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/DotNetCore-SharedPackageStore.md) for more info.

---
In reply to: [110770450](https://github.com/dotnet/corefx/pull/18035#discussion_r110770450) [](ancestors = 110770450)

weshaggard commented on init-tools.cmd:
```diff
@@ -10,8 +10,8 @@ if [%BUILDTOOLS_SOURCE%]==[] set BUILDTOOLS_SOURCE=https://dotnet.myget.org/F/do
 set /P BUILDTOOLS_VERSION=< "%~dp0BuildToolsVersion.txt"
 set BUILD_TOOLS_PATH=%PACKAGES_DIR%Microsoft.DotNet.BuildTools\%BUILDTOOLS_VERSION%\lib\
 set PROJECT_JSON_PATH=%TOOLRUNTIME_DIR%\%BUILDTOOLS_VERSION%
-set PROJECT_JSON_FILE=%PROJECT_JSON_PATH%\project.json
```
I'd say it is worth renaming these "PROJECT_JSON" variables.

mellinoe commented:
I'm going to try to merge this in as-is and follow up with some more cleanup, so that I can stop dealing with merge conflicts with this (the auto-updater keeps giving me merge conflicts over and over).

Will wait for green.

weshaggard commented:
>@karajas Can you look at the latest commit? I added some logic to GenerateDepsFile which filters out items from the "native" section of the file when they aren't contained in our test folder. I was hitting errors because on Windows, the deps file listed "DiaSymReader.dll" when we don't have that in our test folder. There were a handful of such files that are filtered out and removed with my new logic.

The easiest solution here is to add a reference to diasymreader package when on windows. That will also help with some of or test run diagnostics.

mellinoe commented:
I've removed the Alpine build definitions from pipeline.json in the latest commit. The builds are already neglected and non-functional; they would take quite a bit more work to bring them up with this.

eerhardt commented on init-tools.sh:
```diff
@@ -144,24 +145,20 @@ if [ ! -e $__INIT_TOOLS_DONE_MARKER ]; then
         echo "Initializing BuildTools..."
         echo "Running: $__BUILD_TOOLS_PATH/init-tools.sh $__scriptpath $__DOTNET_CMD $__TOOLRUNTIME_DIR" >> $__init_tools_log
+
+        chmod +x $__BUILD_TOOLS_PATH/init-tools.sh
```
I think these "chmod" hacks should have links to the underlying issue:  [https://github.com/NuGet/Home/issues/4424](https://github.com/NuGet/Home/issues/4424)

mellinoe replied:
I'll put that on my list of follow-up fixes
